### PR TITLE
[Logan] Performance improvements for the diffing algorithm.

### PR DESCRIPTION
**Pull request by ian97531**
With this change, I'm seeing 50x - 100x improvements in overall diffing performance. Documents that previously took ~26seconds now take ~300ms.

I kept the basic algorithm the same, but changed it from recursive to iterative, and added a bunch of little optimizations.

### CODE REVIEW

File: js/htmldiff.js [hunk 1]
```diff
@@ -109,12 +123,19 @@
    * @param {number} start_in_after The index of the first token in the list of after tokens.
    * @param {number} length The number of consecutive matching tokens in this block.
    */
-  function Match(start_in_before, start_in_after, length){
-    this.start_in_before = start_in_before;
-    this.start_in_after = start_in_after;
+  function Match(start_in_before, start_in_after, length, segment){
```
Comment (loganfsmyth) [inline js/htmldiff.js]: Needs JSDocs for `segment`.

File: js/htmldiff.js [hunk 2]
```diff
@@ -109,12 +123,19 @@
...
-    this.end_in_before = (this.start_in_before + this.length) - 1;
-    this.end_in_after = (this.start_in_after + this.length) - 1;
+    
+    this.segment_start_in_before = start_in_before;
+    this.segment_start_in_after = start_in_after;
+    this.segment_end_in_before = (this.segment_start_in_before + this.length) - 1;
+    this.segment_end_in_after = (this.segment_start_in_after + this.length) - 1;
+
+    this.end_in_before = segment.before_index + this.segment_end_in_before;
+    this.end_in_after = segment.after_index + this.segment_end_in_after;
```
Comment (loganfsmyth) [inline js/htmldiff.js]: These were hard for me to process. Think it'd be clearer as

    this.end_in_before = this.start_in_before + this.length - 1;
    this.end_in_after = this. start_in_after + this.length - 1;

File: js/htmldiff.js [hunk 3]
```diff
@@ -437,7 +680,8 @@
     for (var i = 0; i < operations.length; i++){
       var op = operations[i];
 
-      if ((is_single_whitespace(op) && last_op.action === 'replace') || (op.action === 'replace' && last_op.action === 'replace')){
+      if ((is_single_whitespace(op) && last_op.action === 'replace') ||
```
Comment (loganfsmyth) [inline js/htmldiff.js]: FYI, looks like this PR has a ton of trailing spaces on lines like this.

**Comment by loganfsmyth** [comment]
Since we're splitting this deploy, I landed my other PR with updates, so this has a conflict now. Should be easy to resolve though.

File: js/htmldiff.js (hunk 1 above)
Comment (ian97531) [inline js/htmldiff.js]: Resolved.

**Comment by ian97531** [comment]
@loganfsmyth I think I've addressed all of your comments. I re-ran the tests and they look good. I also verified by diffing some real content and it looks good as well.

**Comment by loganfsmyth** [comment]
LGTM

**Pull request merged**